Userland/Libraries/LibJS/Runtime/TypedArrayPrototype.cpp

```diff
@@ -1127,146 +1322,156 @@ JS_DEFINE_NATIVE_FUNCTION(TypedArrayPrototype::reverse)
 
-    // 9. Let srcType be TypedArrayElementType(source).
-    // 10. Let srcElementSize be TypedArrayElementSize(source).
+    // 12. Let srcType be TypedArrayElementType(source).
+    // 13. Let srcElementSize be TypedArrayElementSize(source).
     auto source_element_size = source.element_size();
 
-    // 11. Let srcLength be source.[[ArrayLength]].
-    auto source_length = source.array_length();
-
-    // 12. Let srcByteOffset be source.[[ByteOffset]].
+    // 14. Let srcByteOffset be source.[[ByteOffset]].
     auto source_byte_offset = source.byte_offset();
 
-    // 13. If targetOffset is +∞, throw a RangeError exception.
-    if (isinf(target_offset))
+    // 15. If targetOffset = +∞, throw a RangeError exception.
+    if (Value { target_offset }.is_positive_infinity())
         return vm.throw_completion<RangeError>(ErrorType::TypedArrayInvalidTargetOffset, "finite");
 
-    // 14. If srcLength + targetOffset > targetLength, throw a RangeError exception.
+    // 16. If srcLength + targetOffset > targetLength, throw a RangeError exception.
     Checked<size_t> checked = source_length;
     checked += static_cast<u32>(target_offset);
     if (checked.has_overflow() || checked.value() > target_length)
         return vm.throw_completion<RangeError>(ErrorType::TypedArrayOverflowOrOutOfBounds, "target length");
 
-    // 15. If target.[[ContentType]] ≠ source.[[ContentType]], throw a TypeError exception.
+    // 17. If target.[[ContentType]] is not source.[[ContentType]], throw a TypeError exception.
     if (target.content_type() != source.content_type())
         return vm.throw_completion<TypeError>(ErrorType::TypedArrayInvalidCopy, target.class_name(), source.class_name());
 
-    // FIXME: 16. If both IsSharedArrayBuffer(srcBuffer) and IsSharedArrayBuffer(targetBuffer) are true, then
-    // FIXME: a. If srcBuffer.[[ArrayBufferData]] and targetBuffer.[[ArrayBufferData]] are the same Shared Data Block values, let same be true; else let same be false.
+    auto same_shared_array_buffer = false;
 
-    // 17. Else, let same be SameValue(srcBuffer, targetBuffer).
-    auto same = same_value(source_buffer, target_buffer);
+    // 18. If IsSharedArrayBuffer(srcBuffer) is true, IsSharedArrayBuffer(targetBuffer) is true, and srcBuffer.[[ArrayBufferData]] is targetBuffer.[[ArrayBufferData]], let sameSharedArrayBuffer be true; otherwise, let sameSharedArrayBuffer be false.
+    if (source_buffer->is_shared_array_buffer() && target_buffer->is_shared_array_buffer() && (&source_buffer->buffer() == &target_buffer->buffer()))
+        same_shared_array_buffer = true;
 
     size_t source_byte_index = 0;
 
-    // 18. If same is true, then
-    if (same) {
-        // a. Let srcByteLength be source.[[ByteLength]].
-        auto source_byte_length = source.byte_length();
+    // 19. If SameValue(srcBuffer, targetBuffer) is true or sameSharedArrayBuffer is true, then
+    if (same_value(source_buffer, target_buffer) || same_shared_array_buffer) {
```

Comment:
nit: I know this matches the order of the spec, and it probably doesn't matter here, but it feels like we should reorder this to just put `same_shared_array_buffer` here before falling back to `SameValue`.

Reply:
Sure